python/beeai_framework/agents/experimental/agent.py

```diff
@@ -98,22 +102,31 @@ def __init__(
         self._tools = list(tools or [])
         self._requirements = list(requirements or [])
         self._meta = AgentMeta(name=name or "", description=description or "", tools=self._tools)
-        self.middlewares.extend(middlewares or [])
 
-    def run(
-        self,
-        prompt: str | None = None,
-        *,
-        context: str | None = None,
-        expected_output: str | type[TOutput] | None = None,
-        execution: AgentExecutionConfig | None = None,
-        signal: AbortSignal | None = None,
-    ) -> Run[RequirementAgentRunOutput[TOutput]]:
-        run_config = execution or AgentExecutionConfig(
-            max_retries_per_step=3,
-            total_max_retries=3,
-            max_iterations=20,
+    @runnable_entry
+    async def run(self, input: str | list[AnyMessage], /, **kwargs: Unpack[AgentOptions]) -> RequirementAgentOutput:
+        """Execute the agent.
+
+        Args:
+            input: The input to the agent (if list of messages, uses the last message as input)
+            expected_output: Pydantic model or instruction for steering the agent towards an expected output format.
+            total_max_retries: Maximum number of model retries.
+            max_retries_per_step: Maximum number of model retries per step.
+            max_iterations: Maximum number of iterations.
+            backstory: Additional piece of information or background for the agent.
+            signal: The agent abort signal
+            context: A dictionary that can be used to pass additional context to the agent
+
+        Returns:
+            The agent output.
+        """
+        text_input = input if isinstance(input, str) else (input[-1].text if input else "")
+        run_config = AgentExecutionConfig(
+            max_retries_per_step=kwargs.get("max_retries_per_step", 3),
+            total_max_retries=kwargs.get("total_max_retries", 3),
+            max_iterations=kwargs.get("max_iterations", 20),
```

Comment:
Idea: Maybe we could move `AgentExecutionConfig` from `types.py` to `base.py` and add a static method like `.from_options(...)` that would do this for us.

Reply:
I like that, and though of something very similar. Let's push this change to the next PR.